File: Fundamentals_of_Software_Design/2_Algorithms/Сортировка слиянием/main.py
from typing import List
import random

def merge_sort(arr: List[int]) -> List[int]:
    """Сортировка слиянием с использованием единственного вспомогательного массива."""
    if len(arr) <= 1:
        return arr

    temp = arr[:]  # Выделяем один вспомогательный массив
    _merge_sort(arr, temp, 0, len(arr))
    return arr

def _merge_sort(arr: List[int], temp: List[int], left: int, right: int):
    """Рекурсивная сортировка слиянием."""
    if right - left <= 1:
        return
    mid = (left + right) // 2
    _merge_sort(arr, temp, left, mid)
    _merge_sort(arr, temp, mid, right)
    merge(arr, temp, left, mid, right)

def merge(arr: List[int], temp: List[int], left: int, mid: int, right: int):
    """Слияние двух отсортированных частей."""
    i, j, k = left, mid, left
    while i < mid and j < right:
        if arr[i] < arr[j]:
            temp[k] = arr[i]
            i += 1
        else:
            temp[k] = arr[j]
            j += 1
        k += 1
    while i < mid:
        temp[k] = arr[i]
        i += 1
        k += 1
    while j < right:
        temp[k] = arr[j]
        j += 1
        k += 1
    arr[left:right] = temp[left:right]

# Тесты
def test():
    assert merge_sort([]) == []
    assert merge_sort([42]) == [42]
    assert merge_sort([5, 2, 8, 3, 1]) == [1, 2, 3, 5, 8]
    assert merge_sort([3, 3, 3, 3]) == [3, 3, 3, 3]
    assert merge_sort([-2, -5, 0, 7, 1]) == [-5, -2, 0, 1, 7]
    print("Все тесты пройдены!")

# Генерация случайных массивов и тестирование
def random_test(num_tests: int, max_size: int, max_value: int):
    for _ in range(num_tests):
        size = random.randint(0, max_size)
        arr = [random.randint(-max_value, max_value) for _ in range(size)]
        assert merge_sort(arr[:]) == sorted(arr)
    print("Все случайные тесты пройдены!")

if __name__ == "__main__":
    test()
    random_test(100, 1000, 1000)

File: Fundamentals_of_Software_Design/2_Algorithms/Сортировка слиянием/main.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <assert.h>

// Функция слияния двух подмассивов
void merge(int *arr, int *temp, int left, int mid, int right) {
    int i = left, j = mid, k = left;
    while (i < mid && j < right) {
        if (arr[i] < arr[j]) {
            temp[k++] = arr[i++];
        } else {
            temp[k++] = arr[j++];
        }
    }
    while (i < mid) temp[k++] = arr[i++];
    while (j < right) temp[k++] = arr[j++];
    
    for (i = left; i < right; i++) {
        arr[i] = temp[i];
    }
}

// Рекурсивная сортировка слиянием
void merge_sort_recursive(int *arr, int *temp, int left, int right) {
    if (right - left <= 1) return;
    int mid = (left + right) / 2;
    merge_sort_recursive(arr, temp, left, mid);
    merge_sort_recursive(arr, temp, mid, right);
    merge(arr, temp, left, mid, right);
}

// Основная функция сортировки
void merge_sort(int *arr, int size) {
    if (size <= 1) return;
    int *temp = (int *)malloc(size * sizeof(int));
    if (!temp) {
        fprintf(stderr, "Ошибка выделения памяти\n");
        exit(EXIT_FAILURE);
    }
    merge_sort_recursive(arr, temp, 0, size);
    free(temp);
}

// Функция для вывода массива
void print_array(int *arr, int size) {
    for (int i = 0; i < size; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

// Функция для тестирования
void test() {
    int arr1[] = {};
    merge_sort(arr1, 0);
    assert(1);  // len=0, просто не падаем

    int arr2[] = {42};
    merge_sort(arr2, 1);
    assert(arr2[0] == 42); // len=1, массив не меняется

    int arr3[] = {5, 2, 8, 3, 1};
    merge_sort(arr3, 5);
    int expected3[] = {1, 2, 3, 5, 8};
    for (int i = 0; i < 5; i++) assert(arr3[i] == expected3[i]);

    int arr4[] = {3, 3, 3, 3};
    merge_sort(arr4, 4);
    assert(arr4[0] == 3 && arr4[1] == 3 && arr4[2] == 3 && arr4[3] == 3);

    printf("Все тесты пройдены!\n");
}

// Генерация случайных массивов и тест
void random_test(int num_tests, int max_size, int max_value) {
    srand(time(NULL));
    for (int t = 0; t < num_tests; t++) {
        int size = rand() % (max_size + 1);
        int *arr = (int *)malloc(size * sizeof(int));
        int *copy = (int *)malloc(size * sizeof(int));

        for (int i = 0; i < size; i++) {
            arr[i] = rand() % (2 * max_value + 1) - max_value;
            copy[i] = arr[i];
        }

        merge_sort(arr, size);
        qsort(copy, size, sizeof(int), (int (*)(const void *, const void *))compare);

        for (int i = 0; i < size; i++) assert(arr[i] == copy[i]);

        free(arr);
        free(copy);
    }
    printf("Все случайные тесты пройдены!\n");
}

// Сравнительная функция для qsort
int compare(const int *a, const int *b) {
    return (*a > *b) - (*a < *b);
}

int main() {
    test();
    random_test(100, 1000, 1000);
    return 0;
}
